### ISSUE TITLE
[iOS] NRE in ContextActionsCell

### ISSUE THREAD

**Issue by melimion** [xamarin/Xamarin.Forms#4267]
### Description
NRE was introduced in https://github.com/xamarin/Xamarin.Forms/pull/3714

_scroller accessed after it has been disposed.

 It happens in two places - Close() and ScrollDelegate

### Steps to Reproduce

1. Create ListView with many cells and context actions
2. Scroll

### Expected Behavior
No crash
### Actual Behavior
Crash with NullReferenceException in 
```
Xamarin.Forms.Platform.iOS.ContextActionsCell.Close()<22793afa32544bc9bd4626e57194b3f0>:0
Xamarin.Forms.Platform.iOS.CellTableViewCell.GetNativeCell(UITableView tableView, Cell cell, bool recycleCells, string templateId)<22793afa32544bc9bd4626e57194b3f0>:0
Xamarin.Forms.Platform.iOS.ListViewRenderer.ListViewDataSource.GetCell(UITableView tableView, NSIndexPath indexPath)<22793afa32544bc9bd4626e57194b3f0>:0
at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
UIKit.UIApplication.Main(string[] args, IntPtr principal, IntPtr delegate)UIApplication.cs:79
UIKit.UIApplication.Main(string[] args, string principalClassName, string delegateClassName)UIApplication.cs:63
```
or in 
```
Xamarin.Forms.Platform.iOS.ContextActionsCell.get_ScrollDelegate()<22793afa32544bc9bd4626e57194b3f0>:0
Xamarin.Forms.Platform.iOS.ContextActionsCell.PrepareForDeselect()<22793afa32544bc9bd4626e57194b3f0>:0
Xamarin.Forms.Platform.iOS.ListViewRenderer.ListViewDataSource.OnItemSelected(object sender, SelectedItemChangedEventArgs eventArg)<22793afa32544bc9bd4626e57194b3f0>:0
at (wrapper delegate-invoke) System.EventHandler`1[Xamarin.Forms.SelectedItemChangedEventArgs].invoke_void_object_TEventArgs(object,Xamarin.Forms.SelectedItemChangedEventArgs)
Xamarin.Forms.ListView.OnSelectedItemChanged(BindableObject bindable, object oldValue, object newValue)
Xamarin.Forms.BindableObject.SetValueActual(BindableProperty property, BindableObject.BindablePropertyContext context, object value, bool currentlyApplying, SetValueFlags attributes, bool silent)
Xamarin.Forms.BindableObject.SetValueCore(BindableProperty property, object value, SetValueFlags attributes, BindableObject.SetValuePrivateFlags privateAttributes)
Xamarin.Forms.BindableObject.SetValueCore(BindableProperty property, object value, SetValueFlags attributes)
Xamarin.Forms.ListView.NotifyRowTapped(int groupIndex, int inGroupIndex, Cell cell)
Xamarin.Forms.Platform.iOS.ListViewRenderer.ListViewDataSource.RowSelected(UITableView tableView, NSIndexPath indexPath)<22793afa32544bc9bd4626e57194b3f0>:0
Xamarin.Forms.Platform.iOS.ContextActionsCell.SelectGestureRecognizer.Tapped(UIGestureRecognizer recognizer)<22793afa32544bc9bd4626e57194b3f0>:0
UIKit.UITapGestureRecognizer.Callback.Activated(UITapGestureRecognizer sender)UIGestureRecognizer.cs:205
at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
UIKit.UIApplication.Main(string[] args, IntPtr principal, IntPtr delegate)UIApplication.cs:79
UIKit.UIApplication.Main(string[] args, string principalClassName, string delegateClassName)UIApplication.cs:63
```
### Basic Information

- Version with issue: 3.3.0.912540
- Last known good version: 3.2.0.839982

- IDE: VS 15.8.5
- Platform Target Frameworks:
  - iOS:  12.0
- Affected Devices: iPhone/iPad

### Reproduction Link

**Comment by rmarinho** [comment]
Is it possible to provide a small reproduction? thanks

**Comment by melimion** [comment]
While trying to create reproduction for first two stack traces, I got third one that is easy to reproduce.
```
  at Xamarin.Forms.Platform.iOS.ContextScrollViewDelegate.GetContextCell (UIKit.UIScrollView scrollView) [0x00006] in <22793afa32544bc9bd4626e57194b3f0>:0 
  at Xamarin.Forms.Platform.iOS.ContextScrollViewDelegate.ClearCloserRecognizer (UIKit.UIScrollView scrollView) [0x00018] in <22793afa32544bc9bd4626e57194b3f0>:0 
  at Xamarin.Forms.Platform.iOS.ContextScrollViewDelegate+<>c__DisplayClass25_0.<WillEndDragging>b__0 () [0x00046] in <22793afa32544bc9bd4626e57194b3f0>:0 
  at UIKit.UIGestureRecognizer+ParameterlessDispatch.Activated () [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.12.0.4/src/Xamarin.iOS/UIKit/UIGestureRecognizer.cs:69 
--- End of stack trace from previous location where exception was thrown ---
  at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.12.0.4/src/Xamarin.iOS/UIKit/UIApplication.cs:79 
  at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x0002c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.12.0.4/src/Xamarin.iOS/UIKit/UIApplication.cs:63 
  at Xf.Sandbox.iOS.Application.Main (System.String[] args) [0x00001] in /Users/gamer/Projects/Xf.Sandbox/Xf.Sandbox.iOS/Main.cs:17 
```
I think it has same reason
### Steps to Reproduce

1. Run sample
2. Open context menu on one of list cells (with index 1 for example)
3. Tap "scroll" toolbar item
4. Tap on any list items (with index 4999 for example)

 [Xf.Sandbox.zip](https://github.com/xamarin/Xamarin.Forms/files/2534795/Xf.Sandbox.zip)

**Comment by melimion** [comment]
And here is reproduction for first stack trace

### Steps to Reproduce

1. Run sample
2. Tap "insert" toolbar item (You may need to tap "insert" several times, because it doesn't crash every time. But usually 2-3 times enough to crash.)

[Xf.Sandbox.zip](https://github.com/xamarin/Xamarin.Forms/files/2534932/Xf.Sandbox.zip)

**Comment by pauldipietro** [comment]
Ran the first reproduction provided and the crash occurred as described.

**Comment by AndreiMisiukevich** [comment]
The test case
```csharp
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using Xamarin.Forms.CustomAttributes;
using Xamarin.Forms.Internals;

#if UITEST
using NUnit.Framework;
using Xamarin.UITest;
#endif

namespace Xamarin.Forms.Controls.Issues
{
	[Preserve(AllMembers = true)]
	[Issue(IssueTracker.Github, 4267, "[iOS] NRE in ContextActionsCells", PlatformAffected.iOS)]
	public class Issue4267 : TestContentPage
	{
		private readonly ObservableCollection<Issue4267CellViewModel> _cells = new ObservableCollection<Issue4267CellViewModel>();

		protected override void Init()
		{
			Init(10000);

			var list = new Issue4267ListView();
			list.ItemsSource = _cells;
			Content = list;

			var ti = new ToolbarItem();
			ti.Text = "scroll";
			ti.Command = new Command(() =>
			{
				list.ScrollTo(_cells[5000], ScrollToPosition.MakeVisible, false);
			});
			ToolbarItems.Add(ti);
		}

		private void Init(int count)
		{
			_cells.Clear();

			var items = Generate(count);
			for (var i = 0; i < items.Count; i++)
			{
				_cells.Add(items[i]);
			}
		}
		private List<Issue4267CellViewModel> Generate(int count)
		{
			var result = new List<Issue4267CellViewModel>();
			var rand = new Random();

			for (var i = 0; i < count; i++)
			{
				var rows = rand.Next(5, 20);
				result.Add(CreateCell(rows, i));
			}
			return result;
		}
		private Issue4267CellViewModel CreateCell(int rows, int index)
		{
			var sb = new StringBuilder();
			for (var j = 0; j < rows; j++)
			{
				sb.AppendFormat("cell #{0}\n", index);
			}

			var item = new Issue4267CellViewModel { Text = sb.ToString().Trim() };
			return item;
		}


		[Preserve(AllMembers = true)]
		private class Issue4267Cell : ViewCell
		{
			public Issue4267Cell()
			{
				var layout = new StackLayout()
				{
					Padding = new Thickness(5)
				};
				var text = new Label()
				{
					BackgroundColor = Color.Silver,
					TextColor = Color.Black,
				};
				text.SetBinding(Label.TextProperty, nameof(Issue4267CellViewModel.Text));
				layout.Children.Add(text);
				View = layout;

				var ca1 = new MenuItem();
				ca1.Text = "action1";
				ContextActions.Add(ca1);
				var ca2 = new MenuItem();
				ca2.Text = "action2";
				ContextActions.Add(ca2);
			}
		}

		private class Issue4267CellViewModel
		{
			public string Text { get; set; }
		}

		private class Issue4267ListView : Xamarin.Forms.ListView
		{
			public Issue4267ListView() : base(ListViewCachingStrategy.RetainElement)
			{
				ItemSelected += (sender, e) => { ((Xamarin.Forms.ListView)sender).SelectedItem = null; };
				IsPullToRefreshEnabled = false;
				SeparatorVisibility = SeparatorVisibility.None;
				HasUnevenRows = true;
				ItemTemplate = new DataTemplate(typeof(Issue4267Cell));
			}
		}
	}
}
```

**Comment by melimion** [comment]
@rmarinho @PureWeen

Any plans on fixing this?

**Comment by codeknox** [comment]
I just got kind of the same stack on my iOS app:
```
System.NullReferenceException: Object reference not set to an instance of an object
  at Xamarin.Forms.Platform.iOS.ContextActionsCell.get_ScrollDelegate () [0x00000] in <0648e2dffe9e4201b8c6e274ced6579f>:0
  at Xamarin.Forms.Platform.iOS.ContextActionsCell.PrepareForDeselect () [0x00000] in <0648e2dffe9e4201b8c6e274ced6579f>:0
  at Xamarin.Forms.Platform.iOS.ListViewRenderer+ListViewDataSource.OnItemSelected (System.Object sender, Xamarin.Forms.SelectedItemChangedEventArgs eventArg) [0x00068] in <0648e2dffe9e4201b8c6e274ced6579f>:0
  at at (wrapper delegate-invoke) System.EventHandler`1[Xamarin.Forms.SelectedItemChangedEventArgs].invoke_void_object_TEventArgs(object,Xamarin.Forms.SelectedItemChangedEventArgs)
  at Xamarin.Forms.ListView.OnSelectedItemChanged (Xamarin.Forms.BindableObject bindable, System.Object oldValue, System.Object newValue) [0x00000] in D:\a\1\s\Xamarin.Forms.Core\ListView.cs:601
  at Xamarin.Forms.BindableObject.SetValueActual (Xamarin.Forms.BindableProperty property, Xamarin.Forms.BindableObject+BindablePropertyContext context, System.Object value, System.Boolean currentlyApplying, Xamarin.Forms.Internals.SetValueFlags attributes, System.Boolean silent) [0x00120] in D:\a\1\s\Xamarin.Forms.Core\BindableObject.cs:624
  at Xamarin.Forms.BindableObject.SetValueCore (Xamarin.Forms.BindableProperty property, System.Object value, Xamarin.Forms.Internals.SetValueFlags attributes, Xamarin.Forms.BindableObject+SetValuePrivateFlags privateAttributes) [0x0019f] in D:\a\1\s\Xamarin.Forms.Core\BindableObject.cs:430
  at Xamarin.Forms.BindableObject.SetValueCore (Xamarin.Forms.BindableProperty property, System.Object value, Xamarin.Forms.Internals.SetValueFlags attributes) [0x00000] in D:\a\1\s\Xamarin.Forms.Core\BindableObject.cs:367
  at Xamarin.Forms.ListView.NotifyRowTapped (System.Int32 groupIndex, System.Int32 inGroupIndex, Xamarin.Forms.Cell cell) [0x0004c] in D:\a\1\s\Xamarin.Forms.Core\ListView.cs:441
  at Xamarin.Forms.Platform.iOS.ListViewRenderer+ListViewDataSource.RowSelected (UIKit.UITableView tableView, Foundation.NSIndexPath indexPath) [0x0006f] in <0648e2dffe9e4201b8c6e274ced6579f>:0
  at Xamarin.Forms.Platform.iOS.ContextActionsCell+SelectGestureRecognizer.Tapped (UIKit.UIGestureRecognizer recognizer) [0x00044] in <0648e2dffe9e4201b8c6e274ced6579f>:0
  at UIKit.UITapGestureRecognizer+Callback.Activated (UIKit.UITapGestureRecognizer sender) [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.8.0.2/src/Xamarin.iOS/UIKit/UIGestureRecognizer.cs:223
  at at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.8.0.2/src/Xamarin.iOS/UIKit/UIApplication.cs:79
  at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x0002c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.8.0.2/src/Xamarin.iOS/UIKit/UIApplication.cs:63
  at SkyWell.iOS.Application.Main (System.String[] args) [0x00026] in /Users/sibagy/fileZ/projZ/SkyWell/src/SkyWell/SkyWell.iOS/Main.cs:13
```

These are the versions I'm using:
```
=== Visual Studio Enterprise 2019 for Mac ===

Version 8.0.5 (build 9)
Installation UUID: 43dd0835-b143-41aa-8695-41efb931ca7a
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 5.6.0.2 (d16-0 / 040682909)

	Package version: 518010003

=== Mono Framework MDK ===

Runtime:
	Mono 5.18.1.3 (2018-08/fdb26b0a445) (64-bit)
	Package version: 518010003

=== NuGet ===

Version: 4.8.2.5835

=== .NET Core ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	2.1.9
	2.1.8
	2.1.7
	2.1.6
	2.1.2
	2.1.1
	2.0.5
SDK: /usr/local/share/dotnet/sdk/2.1.505/Sdks
SDK Versions:
	2.1.505
	2.1.504
	2.1.503
	2.1.500
	2.1.302
	2.1.301
	2.1.4
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/5.18.1/lib/mono/msbuild/15.0/bin/Sdks

=== Xamarin.Profiler ===

Version: 1.6.9
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Updater ===

Version: 11

=== Xamarin.Android ===

Version: 9.2.3.0 (Visual Studio Enterprise)
Android SDK: /Users/sibagy/Library/Android/sdk
	Supported Android versions:
		6.0 (API level 23)
		7.0 (API level 24)
		7.1 (API level 25)
		8.0 (API level 26)
		8.1 (API level 27)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 28.0.2
SDK Build Tools Version: 28.0.3

Build Information: 
Mono: mono/mono/2018-08-rc@5ac37ccd385
Java.Interop: xamarin/java.interop/d16-0@c987483
LibZipSharp: grendello/LibZipSharp/d16-1@44de300
LibZip: nih-at/libzip/rel-1-5-1@b95cf3f
MXE: xamarin/mxe/xamarin@b9cbb535
ProGuard: xamarin/proguard/master@905836d
SQLite: xamarin/sqlite/3.26.0@325e91a
Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-0@0a7edd6

=== Microsoft Mobile OpenJDK ===

Java SDK: /Users/sibagy/Library/Developer/Xamarin/jdk/microsoft_dist_openjdk_8.0.25
1.8.0-25
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Android Device Manager ===

Version: 1.2.0.14
Hash: 86df26f
Branch: remotes/origin/d16-0
Build date: 2019-04-05 16:58:56 UTC

=== Xamarin Designer ===

Version: 4.17.4.418
Hash: 3d086e814
Branch: remotes/origin/d16-0
Build date: 2019-04-01 09:20:10 UTC

=== Apple Developer Tools ===

Xcode 10.2.1 (14490.122)
Build 10E1001

=== Xamarin.Mac ===

Version: 5.8.0.0 (Visual Studio Enterprise)
Hash: 0aa84521
Branch: d16-0
Build date: 2019-04-02 16:01:19-0400

=== Xamarin.iOS ===

Version: 12.8.0.2 (Visual Studio Enterprise)
Hash: f2248ae6
Branch: d16-0
Build date: 2019-04-23 11:59:04-0400

=== Xamarin Inspector ===

Version: 1.4.3
Hash: db27525
Branch: 1.4-release
Build date: Mon, 09 Jul 2018 21:20:18 GMT
Client compatibility: 1

=== Build Information ===

Release ID: 800050009
Git revision: d9d7eed60cb59740ba55c04bfe9c39ac57014c12
Build date: 2019-04-26 10:08:19+00
Build branch: master
Xamarin extensions: 5efedb990d786142d3b54351c01ce867af7b7680

=== Operating System ===

Mac OS X 10.14.4
Darwin 18.5.0 Darwin Kernel Version 18.5.0
    Mon Mar 11 20:40:32 PDT 2019
    root:xnu-4903.251.3~3/RELEASE_X86_64 x86_64

```

Any idea of RTM of the fix?

**Comment by sermetk** [comment]
@melimion In my case SelectedItem = null causing this error, I commented out temporarily.

**Comment by v-haroha** [comment]
I was testing this bug out on the latest Xamarin.Forms project, but using the code for the provided example (Xf.Sandbox.zip) and got the following error while spamming the "Insert" button

{System.NullReferenceException: Object reference not set to an instance of an object
  at Xamarin.Forms.Platform.iOS.ContextActionsCell.Close () [0x00000] in D:\a\1\s\Xamarin.Forms.Platform.iOS\ContextActionCell.cs:84 
  at Xamarin.Forms.Platform.iOS.CellTableViewCell.GetNativeCell (UIKit.UITableView tableView, Xamarin.Forms.Cell cell, System.Boolean recycleCells, System.String templateId) [0x00055] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Cells\CellTableViewCell.cs:64 
  at Xamarin.Forms.Platform.iOS.ListViewRenderer+ListViewDataSource.GetCell (UIKit.UITableView tableView, Foundation.NSIndexPath indexPath) [0x00029] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ListViewRenderer.cs:1013 
  at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.14.0.110/src/Xamarin.iOS/UIKit/UIApplication.cs:86 
  at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x0000e] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.14.0.110/src/Xamarin.iOS/UIKit/UIApplication.cs:65 
  at TestScrollTooBug.iOS.Application.Main (System.String[] args) [0x00001] in C:\Users\v-haroha\source\repos\TestScrollTooBug\TestScrollTooBug.iOS\Main.cs:17 }

**Comment by v-haroha** [comment]
I think I found around what causes the issue, but I would need to defer to the Xamarin Dev team for the expertise on why this is doing what is it doing.
Under  Xamarin.Forms.Platform.iOS.ContextActionCell.cs line 64 under method GetNativeCell a call is made to close the contextCell, but if the ListView is setup to use RetainElement, then I guess this causes issues. I guess this is the case because it isnt setup to close Cells in RetainElement.

Now the reason it is running this code the example is because of attached ContextActions. The cells are setup with context Actions line 54 of ContextActionCell has an if statement that says
	`if (cell.HasContextActions || recycleCells)`
So with this IF a RetainElement cell is treated as a RecycleElement, and thus crashing the listview and the app.

I'm not sure on how to fix this without knowing more of the "why's", but I have found that setting your ListView to RecycleElement does resolve the issue
